src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs

```diff
@@ -2632,22 +2632,81 @@ protected override void VisitBinaryOperatorChildren(ArrayBuilder<BoundBinaryOper
             var binary = stack.Pop();
 
             var (leftOperand, leftConversion) = RemoveConversion(binary.Left, includeExplicitConversions: false);
-            VisitRvalue(leftOperand);
+            Visit(leftOperand);
 
             while (true)
             {
-                AfterLeftChildHasBeenVisited(leftOperand, leftConversion, binary);
+                if (!learnFromBooleanConstantTest())
+                {
+                    Unsplit(); // VisitRvalue does this
+                    UseRvalueOnly(leftOperand); // drop lvalue part
+
+                    AfterLeftChildHasBeenVisited(leftOperand, leftConversion, binary);
+                }
 
                 if (stack.Count == 0)
                 {
                     break;
                 }
 
-                Unsplit(); // VisitRvalue does this
                 leftOperand = binary;
                 leftConversion = Conversion.Identity;
                 binary = stack.Pop();
             }
+
+            // learn from true/false constant
+            bool learnFromBooleanConstantTest()
```

Comment:
>learnFromBooleanConstantTest [](start = 17, length = 28)

Is there a reason to do this work here rather than in `AfterLeftChildHasBeenVisited()` where the rest of the analysis is?

Reply:
`AfterLeftChildHasBeenVisited` doesn't work in conditional state (it contains multiple assertions).